### ISSUE TITLE
Making a key required while extending a dictionary validation schema fails

### ISSUE THREAD

**Issue by balloob** [alecthomas/voluptuous#192]
Running voluptuous 0.9.2

Extending a dictionary validating schema will not work if one of the existing keys is overwritten and made required.

Given this code, you would expect both schema's to pass the data:

```python
import voluptuous as vol

schema1 = vol.Schema({'platform': str})
schema2 = schema1.extend({vol.Required('platform'): str})

data = {'platform': 'hello'}

print('schema1')
print(schema1(data))
print('schema2')
print(schema2(data))
```

However, an invalid exception is raised for `schema2`:

```
› python3 pm_test.py
schema1
{'platform': 'hello'}
schema2
Traceback (most recent call last):
  File "pm_test.py", line 20, in <module>
    print(schema2(data))
  File "/Users/paulus/dev/python/home-assistant/lib/python3.4/site-packages/voluptuous-0.9.2-py3.4.egg/voluptuous/schema_builder.py", line 192, in __call__
  File "/Users/paulus/dev/python/home-assistant/lib/python3.4/site-packages/voluptuous-0.9.2-py3.4.egg/voluptuous/schema_builder.py", line 486, in validate_dict
  File "/Users/paulus/dev/python/home-assistant/lib/python3.4/site-packages/voluptuous-0.9.2-py3.4.egg/voluptuous/schema_builder.py", line 324, in validate_mapping
voluptuous.error.MultipleInvalid: required key not provided @ data['platform']
```

**Comment by sfelf** [comment]
This is the same behavior if you declare both the Optional and the Required in the initial Schema.

```
>>> from voluptuous import Required, Schema
>>> s = Schema({'test': str, Required('test'): str}, required=False)
>>> s({'test': 'this'})
Traceback (most recent call last):
  File "<input>", line 3, in <module>
  File "/Users/test/virtual/test/lib/python3.4/site-packages/voluptuous.py", line 337, in __call__ 
    return self._compiled([], data)
  File "/Users/test/virtual/test/lib/python3.4/site-packages/voluptuous.py", line 635, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/Users/test/virtual/test/lib/python3.4/site-packages/voluptuous.py", line 471, in validate_mapping
    raise MultipleInvalid(errors)
voluptuous.MultipleInvalid: required key not provided @ data['test']
```

**Comment by tusharmakkar08** [comment]
Closed via 99fb7bd43b2cbd8e53cb9e50765c93339b2f7843
